.storybook/addons/withSmartKnobs.js

```diff
@@ -0,0 +1,182 @@
+// Default simple PropType based knob map.
+const propTypeKnobsMap = [
+  { name: 'string', knob: text },
+  { name: 'number', knob: number },
+  { name: 'bool', knob: boolean },
+  { name: 'func', regexp: /=>/, knob: (name, value) => value || action(name) },
+  { name: 'object', regexp: /^{.*}$/, knob: object },
+  { name: 'node', regexp: /^ReactNode$/, knob: text },
+  { name: 'element', knob: text },
+  { name: 'array', regexp: /\[]$|^\[.*]$/, knob: object },
+]
+
+const typeKnobsMap = [...flowTypeKnobsMap, ...propTypeKnobsMap]
+
+typeKnobsMap.forEach(({ name, regexp, knob, args = [] }, weight) => addKnobResolver({
+  weight: weight * 10,
+  name: `PropTypes.${name}`,
+  resolver: propTypeKnobResolver(name, regexp, knob, ...args)
+}))
+
+const optionsReducer = (res, value) => ({ ...res, [value]: value })
+const withDefaultOption = (options) => ({ '--': undefined, ...options })
+const createSelect = (propName, elements, defaultValue, isRequired) => {
+  try {
+    const options = elements
+    // Cleanup string quotes, if any.
+      .map(value => cleanupValue(value.value))
+      .reduce(optionsReducer, {})
+    const value = defaultValue || (isRequired && Object.values(options)[0]) || undefined
+    return select(propName, isRequired ? options : withDefaultOption(options), value)
+  }
+  catch (e) { }
+}
+
+// Register 'oneOf' PropType knob resolver.
+addKnobResolver({
+  name: 'PropTypes.oneOf',
+  resolver: (propName, propType, defaultValue) => {
+    if (propType.type.name === 'enum' && propType.type.value.length) {
+      return createSelect(propName, propType.type.value, defaultValue, propType.required)
+    }
+    // for flow support
+    if (propType.type.name === 'union' && propType.type.elements) {
+      return createSelect(propName, propType.type.elements, defaultValue, propType.required)
+    }
+  }
+})
+
+const ensureType = (item) => item.flowType ? ({ ...item, type: item.flowType }) : item
+
+const getNewProps = (target, context, opts) => {
+  const { __docgenInfo: { props } = { props: {} } } = target.type
+  const { ignoreProps = [] } = opts
+  const defaultProps = {
+    ...(target.type.defaultProps || {}),
+    ...target.props
+  }
+
+  const finalProps = props ? Object.keys(props).reduce((acc, n) => {
+    const item = ensureType(props[n])
+
+    if (ignoreProps.includes(n)) {
+      return acc
+    }
+
+    if (!item.type) {
+      const defaultValue = item.defaultValue ? item.defaultValue.value : 'Unknown'
+      logger.warn(`There is a prop with defaultValue ${defaultValue} but it wasn't specified on element.propTypes. Check story: "${context.kind}".`)
+      return acc
+    }
+
+    return {
+      ...acc,
+      [n]: item,
+    }
+  }, {}) : {}
+
+  return resolvePropValues(finalProps, defaultProps)
+}
+
+const mutateChildren = (component, context, opts) => {
+  return cloneElement(component, { children: Children.map(component.props.children, (child) => {
+    if (child.type && child.type.__docgenInfo) {
```

Comment:
Just out of my own curiosity - what is docgenInfo?

Reply:
It's created by [this Babel plugin](https://github.com/storybookjs/babel-plugin-react-docgen). AFAIK it's an object representation of the components prop types that can be parsed and rendered by Storybook.